lib/plenario2/actions/data_set_diff_actions.ex

```diff
@@ -6,15 +6,45 @@ defmodule Plenario2.Actions.DataSetDiffActions do
 
   import Ecto.Query
 
+  import Plenario2.Guards, only: [is_id: 1]
+
   alias Plenario2.Changesets.DataSetDiffChangesets
-  alias Plenario2.Schemas.{DataSetDiff, Meta}
+  alias Plenario2.Schemas.{DataSetDiff, Meta, DataSetConstraint, EtlJob}
   alias Plenario2.Repo
 
+  @typedoc """
+  Parameter is an ID attribute
+  """
+  @type id :: String.t | integer
+
+  @typedoc """
+  Returns a tuple of :ok, DataSetDiff or :error, Ecto.Changeset
+  """
+  @type ok_diff :: {:ok, DataSetDiff} | {:error, Ecto.Changeset.T}
+
   @doc """
   Creates a new instance of DataSetDiff.
   """
-  @spec create(meta_id :: integer, constraint_id :: integer, etl_job_id :: integer, column :: String.t, original :: any, updated :: any, changed_on :: %DateTime{}, constraint_values :: %{}) :: {:ok, %DataSetDiff{} | :error, Ecto.Changeset.t}
-  def create(meta_id, constraint_id, etl_job_id, column, original, updated, changed_on, constraint_values) do
+  @spec create(meta :: Meta | id, constraint :: DataSetConstraint | id, etl_job :: EtlJob | id, column :: String.t, original :: any, updated :: any, changed_on :: DateTime, constraint_values :: %{}) :: ok_diff
+  def create(meta, constraint, etl_job, column, original, updated, changed_on, constraint_values) do
+    meta_id =
+      case is_id(meta) do
+        true -> meta
+        false -> meta.id
+      end
+
+    constraint_id =
+      case is_id(constraint) do
+        true -> constraint
+        false -> constraint.id
+      end
+
+    etl_job_id =
+      case is_id(etl_job) do
+        true -> etl_job
+        false -> etl_job.id
+      end
```

Comment:
Same thing, just add more heads and they'll eventually call the correct one. As long as the call to create is the last piece of work done, we don't build out a recursive call stack.

```elixir
def create(meta, constraint, job, fields) when !is_id(meta), do: create(meta.id, constraint, job, fields)
def create(meta, constraint, job, fields) when !is_id(constraint), do: create(meta, constraint.id, job fields)
def create(meta, constraint, job, fields) when !is_id(job), do: create(meta, constraint, job.id, fields)
def create(meta_id, fields) do
  # work with ids
end
```